core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/UnwrapCastInComparison.java

```diff
@@ -173,6 +176,33 @@ public Expression rewriteComparisonExpression(ComparisonExpression node, Void co
             return unwrapCast(expression);
         }
 
+        @Override
+        public Expression rewriteInPredicate(InPredicate node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+        {
+            InPredicate inPredicate = treeRewriter.defaultRewrite(node, null);
+            Expression value = inPredicate.getValue();
+            Expression valueList = inPredicate.getValueList();
+
+            if (!(value instanceof Cast) ||
+                    !(valueList instanceof InListExpression inListExpression)) {
+                return inPredicate;
+            }
+
+            // Convert each value to a comparison expression and try to unwrap it.
+            // unwrap the InPredicate only in case we manage to unwrap the entire value list
+            ImmutableList.Builder<Expression> comparisonExpressions = ImmutableList.builderWithExpectedSize(inListExpression.getValues().size());
+            for (Expression rightExpression : inListExpression.getValues()) {
+                ComparisonExpression comparisonExpression = new ComparisonExpression(EQUAL, value, rightExpression);
+                Expression unwrappedExpression = unwrapCast(comparisonExpression);
```

Comment:
Instead of creating a comparison expression, change `unwrapCast` to take the three components of a comparison expression (operation, left, right) and call it directly: `unwrapCast(EQUAL, value, rightExpression)`

Reply:
implemented it in a different commit, had to change the return type of `unwrapCast` to `Optional<Expression>` because in this code I need to know if the unwrap happened or not, so it turned out to be more touched lines than I thought...
take a look. 
had to rebase to get the already merged changes to `UnwrapYear...` aligned with the proposed change